### PR TITLE
Improve host scoring by including MaxCollateral and RemainingStorage

### PR DESCRIPTION
**Pull request by ChrisSchinnerl**
This PR adds gouging checks for `MaxCollateral`, adds a score adjustment mechanism for remaining storage on hosts and extends the contract maintenance with a score check against a minimal score threshold.

That way we should be a lot less likely to select hosts with insufficient storage for contract formation and we will be able to avoid hosts which are not willing to put a minimum amount of collateral into contracts.

### CODE REVIEW

File: autopilot/contractor.go [hunk 1]
```diff
@@ -662,31 +702,30 @@ func (c *contractor) candidateHosts(ctx context.Context, cfg api.AutopilotConfig
...
+		if _, exclude := exclude[h.PublicKey]; exclude {
 			continue
 		}
-		if usable, _ := isUsableHost(cfg, gs, rs, ipFilter, h); !usable {
+		if usable, _ := isUsableHost(cfg, gs, rs, ipFilter, h, minScore, storedData[h.PublicKey]); !usable {
 			continue
 		}
 
-		score := hostScore(cfg, h)
+		score := hostScore(cfg, h, 0, float64(rs.TotalShards)/float64(rs.MinShards))
```
Comment (peterjan) [inline autopilot/contractor.go:719]: `nit` should we just add a `Redundancy() float64` method on the settings?

File: autopilot/hostscore.go [hunk 2]
```diff
@@ -29,16 +29,47 @@ const (
...
+	idealDataPerHost := float64(cfg.Contracts.Storage) * expectedRedundancy / float64(cfg.Contracts.Amount)
+	// allocationPerHost is the amount of data that we would like to be able to
+	// put on each host, because data is not always spread evenly across the
+	// hosts during upload. Slower hosts may get very little data, more
+	// expensive hosts may get very little data, and other factors can skew the
+	// distribution. allocationPerHost takes into account the skew and tries to
+	// ensure that there's enough allocation per host to accommodate for a skew.
+	// NOTE: assume that data is not spread evenly and the host with the most
+	// data will store twice the expectation
+	allocationPerHost := idealDataPerHost * 2
```
Comment (peterjan) [inline autopilot/hostscore.go:55]: `nit` is x2 enough?

File: autopilot/hostscore.go [hunk 3]
```diff
@@ -29,16 +29,47 @@ const (
...
 
-func hostScore(cfg api.AutopilotConfig, h hostdb.Host) float64 {
+func hostScore(cfg api.AutopilotConfig, h hostdb.Host, storedData uint64, expectedRedundancy float64) float64 {
 	// TODO: priceAdjustmentScore
-	// TODO: storageRemainingScore
 	return ageScore(h) *
-		collateralScore(cfg, *h.Settings) *
+		collateralScore(cfg, *h.Settings, expectedRedundancy) *
 		interactionScore(h) *
+		storageRemainingScore(cfg, *h.Settings, 0, 3) *
```
Comment (peterjan) [inline autopilot/hostscore.go:37]: you're not passing `storedData` and `expectedRedundancy` here

File: autopilot/contractor.go [hunk 4]
```diff
@@ -633,7 +646,34 @@ func (c *contractor) renewFundingEstimate(ctx context.Context, cfg api.Autopilot
...
+	if err != nil {
+		return 0, err
+	}
+	if len(hosts) == 0 {
+		return 0, errors.New("no hosts returned in candidateHosts")
+	}
+
+	// Find the minimum score that a host is allowed to have to be considered
+	// good for upload.
+	lowestScore := hostScore(cfg, hosts[0], 0, redundancy)
```
Comment (peterjan) [inline autopilot/contractor.go:666]: `nit` just set `lowestScore` to max float?

File: autopilot/contractor.go [hunk 5]
```diff
@@ -633,7 +646,34 @@ func (c *contractor) renewFundingEstimate(ctx context.Context, cfg api.Autopilot
...
+	for i := 1; i < len(hosts); i++ {
+		score := hostScore(cfg, hosts[i], 0, redundancy)
+		if score < lowestScore {
+			lowestScore = score
+		}
+	}
+	return lowestScore / 500, nil
```
Comment (peterjan) [inline autopilot/contractor.go:673]: where's the `/ 500` coming from?

File: autopilot/contractor.go [hunk 6]
```diff
@@ -633,7 +646,34 @@ func (c *contractor) renewFundingEstimate(ctx context.Context, cfg api.Autopilot
...
 }
 
-func (c *contractor) candidateHosts(ctx context.Context, cfg api.AutopilotConfig, used map[types.PublicKey]bool, wanted int) ([]hostdb.Host, error) {
+func (c *contractor) managedFindMinAllowedHostScores(ctx context.Context, cfg api.AutopilotConfig, storedData map[types.PublicKey]uint64, redundancy float64) (float64, error) {
+	// Pull a new set of hosts from the hostdb that could be used as a new set
+	// to match the allowance. The lowest scoring host of these new hosts will
+	// be used as a baseline for determining whether our existing contracts are
+	// worthwhile.
+	numContracts := cfg.Contracts.Amount
+	buffer := 50
```
Comment (peterjan) [inline autopilot/contractor.go:655]: `nit` not that I really care but why  not reuse `leewayPctCandidateHosts`?

File: autopilot/contractor.go [hunk 7]
```diff
@@ -633,7 +646,34 @@ func (c *contractor) renewFundingEstimate(ctx context.Context, cfg api.Autopilot
...
+	hosts, err := c.candidateHosts(ctx, cfg, make(map[types.PublicKey]struct{}), storedData, int(numContracts)+int(buffer), 1) // 1 to avoid 0 score hosts
```
Comment (peterjan) [inline autopilot/contractor.go:656]: `candidateHosts` is really expensive and we call it twice now (potentially) - kind of unfortunate that every iteration fetches all hosts from the database twice... can we avoid it by keeping track of host scores? So essentially every host we ever encounter contributes to the lowest host score, and we don't calculate `managedFindMinAllowedHostScores` explicitly?

File: autopilot/contractor.go (hunk 6 above)
Comment (ChrisSchinnerl) [reply autopilot/contractor.go:655]: That's a lot lower than 50 for most settings.

File: autopilot/contractor.go (hunk 5 above)
Comment (ChrisSchinnerl) [reply autopilot/contractor.go:673]: updated

File: autopilot/hostscore.go (hunk 2 above)
Comment (ChrisSchinnerl) [reply autopilot/hostscore.go:55]: I think 2x is pretty safe.

File: autopilot/contractor.go (hunk 7 above)
Comment (ChrisSchinnerl) [reply autopilot/contractor.go:656]: I changed it to fetch the hosts only once. That's a bit simpler. With your new pruning code things should become a lot lightweight as well.

File: autopilot/contractor.go (hunk 6 above)
Comment (peterjan) [reply autopilot/contractor.go:666]: I know but where does the 50 come from?

File: autopilot/contractor.go (hunk 7 above)
Comment (peterjan) [reply autopilot/contractor.go]: agreed, I just think we shouldn't necessarily rely on pruning and ensure we don't do this twice - thanks for updating it

**Pull request merged**